### PR TITLE
Start of RTL to LLHD conversion pass

### PR DESCRIPTION
**Pull request by mikeurbach**
@fabianschuiki this series of commits starts a dialect conversion from RTL to LLHD, whenever you are able to take a look. Initially it supports mapping `rtl.module` to `llhd.entity`. I think next would be to add support for `rtl.instance`, wires and connects, and the various combinational ops.

### CODE REVIEW

File: include/circt/Dialect/RTL/Structure.td [hunk 1]
```diff
@@ -32,6 +32,12 @@ def RTLModuleOp : RTLOp<"module",
     // squatting on the name.
     Block *getBodyBlock() { return &body().front(); }
 
+    // Get the module's symbolic name.
+    StringRef getName() {
```
Comment (lattner) [inline include/circt/Dialect/RTL/Structure.td:36]: Feel free to land things like this as an independent patch if it is helpful.

File: lib/Conversion/RTLToLLHD/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+add_circt_conversion_library(MLIRRTLToLLHD
```
Comment (lattner) [inline lib/Conversion/RTLToLLHD/CMakeLists.txt:1]: Plz add file header boilerplate.

File: lib/Conversion/RTLToLLHD/RTLToLLHD.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
+//===- RTLToLLHD.cpp - RTL to LLHD Conversion Pass ------------------------===//
```
Comment (lattner) [inline lib/Conversion/RTLToLLHD/RTLToLLHD.cpp:1]: The code base is inconsistent, but please include this in the file headers:

```
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
```

**Review by lattner — APPROVED** [review]
This looks structurally reasonable to me, but I'm not a domain expert on LLHD

File: lib/Conversion/RTLToLLHD/RTLToLLHD.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  if (failed(applyPartialConversion(module, target, std::move(patterns))))
+    signalPassFailure();
+}
+
+//===----------------------------------------------------------------------===//
+// TypeConverter conversions and materializations
+//===----------------------------------------------------------------------===//
+RTLToLLHDTypeConverter::RTLToLLHDTypeConverter() {
+  // Convert IntegerType by just wrapping it in SigType.
+  addConversion([](IntegerType type) { return SigType::get(type); });
```
Comment (fabianschuiki) [inline lib/Conversion/RTLToLLHD/RTLToLLHD.cpp:84]: This is good start in my opinion. It's not quite clear to me yet how we'll decide to convert something in the RTL dialect to a signal with associated `prb` and `drv`, and when to chain multiple RTL ops together without writing back to a signal. Maybe there is a sort of "signal-to-reg" pass hidden in there somewhere, alongside LLVM's mem-to-reg 🤔.

File: lib/Conversion/RTLToLLHD/RTLToLLHD.cpp [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                                entityBodyRegion.end());
+
+    // Attempt to convert the entry block's args using the previous conversion.
+    if (failed(rewriter.convertRegionTypes(&entityBodyRegion, *typeConverter,
+                                           &argConversion)))
+      return module.emitError("could not convert region types");
+
+    // Add the LLHD terminator op after the RTL module's output ops.
+    rewriter.setInsertionPointToEnd(entity.getBodyBlock());
+    rewriter.create<llhd::TerminatorOp>(entity.getLoc());
```
Comment (fabianschuiki) [inline lib/Conversion/RTLToLLHD/RTLToLLHD.cpp:125]: This is super nice. I really like all the infrastructure that MLIR keeps around for this kind of task. Very neat 😄

File: lib/Dialect/LLHD/IR/LLHDOps.cpp [hunk 6]
```diff
@@ -663,22 +663,30 @@ static LogicalResult verify(llhd::EntityOp op) {
                "Cannot have more inputs than arguments, expected at most ")
            << numArgs << " but got: " << nIns;
   }
+
+  // Check that all block arguments are of signal type
+  for (size_t i = 0; i < numArgs; ++i)
+    if (!op.getArgument(i).getType().isa<llhd::SigType>())
+      return op.emitError("usage of invalid argument type. Got ")
+             << op.getArgument(i).getType() << ", expected LLHD signal type";
+
```
Comment (fabianschuiki) [inline lib/Dialect/LLHD/IR/LLHDOps.cpp:672]: Very nice addition, thanks! :+1:

File: lib/Dialect/LLHD/IR/LLHDOps.cpp [hunk 7]
```diff
@@ -663,22 +663,30 @@ static LogicalResult verify(llhd::EntityOp op) {
...
-  for (int i = 0, e = getNumFuncArguments(); i < e; ++i) {
-    if (!getArgument(i).getType().isa<llhd::SigType>()) {
+  for (Type inputType : type.getInputs())
+    if (!inputType.isa<llhd::SigType>())
       return emitOpError("usage of invalid argument type. Got ")
-             << getArgument(i).getType() << ", expected LLHD signal type";
-    }
-  }
+             << inputType << ", expected LLHD signal type";
+
```
Comment (fabianschuiki) [inline lib/Dialect/LLHD/IR/LLHDOps.cpp:689]: Much cleaner :+1:. I still wonder if at some point we'd want to change LLHD entities to not partition inputs from outputs by a fixed offset, but rather just having one plain argument list with directionality annotated as an attribute. That could make linking much easier, especially if we make port connections by name.

File: test/Conversion/RTLToLLHD/structure.mlir [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  // CHECK-LABEL: llhd.entity @test
+  // CHECK-SAME: (%[[IN:.+]] : !llhd.sig<i1>) ->
+  // CHECK-SAME: (%[[OUT:.+]] : !llhd.sig<i1>)
+  rtl.module @test(%in: i1) -> (%out: i1) {
+    // CHECK: %[[DELTA:.*]] = llhd.const #llhd.time<0ns, 1d, 0e>
+    // CHECK: %[[PRB:.*]] = llhd.prb %[[IN]]
+    // CHECK: llhd.drv %[[OUT]], %[[PRB]] after %[[DELTA]]
+    rtl.output %in: i1
+  }
+}
```
Comment (fabianschuiki) [inline test/Conversion/RTLToLLHD/structure.mlir:13]: Cool 😎

**Review by fabianschuiki — APPROVED** [review]
Looks very nice :+1: Thanks a lot for tackling this! I agree that a good next step would be instantiations, wiring, and some basic combinational operations.

File: lib/Conversion/RTLToLLHD/RTLToLLHD.cpp (hunk 4 above)
Comment (mikeurbach) [reply lib/Conversion/RTLToLLHD/RTLToLLHD.cpp:84]: If we can frame the problem as type conversion, the dialect conversion infrastructure provides some nice helpers to materialize casts. In this case, we could materialize a signal with associated `prb` and `drv` ops as the casts.

I'm not sure how far we can take that with respect to chaining together multiple ops without writing back to a signal, but the cast materialization approach works nicely if it is a good fit.

File: lib/Dialect/LLHD/IR/LLHDOps.cpp (hunk 7 above)
Comment (mikeurbach) [reply lib/Dialect/LLHD/IR/LLHDOps.cpp:689]: That would make sense to me, and might make processing like this a little more simple. I'll leave this as future work for now.

File: include/circt/Dialect/RTL/Structure.td (hunk 1 above)
Comment (mikeurbach) [reply include/circt/Dialect/RTL/Structure.td:36]: :+1: will do. In this case it all kind of flowed together, but I will try to make some small pre-emptive additions like this to cut down on review complexity.

**Pull request merged**